examples/step-x/step-x.cc

```diff
@@ -0,0 +1,542 @@
+    particle_triangulation.refine_global(par.particle_insertion_refinement);
+
+    DoFHandler<dim> particles_dof_handler(particle_triangulation);
+    FE_Q<dim>       particles_fe(1);
+
+    particles_dof_handler.distribute_dofs(particles_fe);
+
+    // Generation of the particles using the Particles::Generators
+    Particles::Generators::dof_support_points(particles_dof_handler,
+                                              global_bounding_boxes,
+                                              particle_handler);
+
+    // Displaying the total number of generated particles in the domain
+    pcout << "Number of particles inserted: "
+          << particle_handler.n_global_particles() << std::endl;
+  }
+
+  // Sets up the background degree of freedom using their interpolation
+  // And allocated a vector where you can store the entire solution
+  // of the velocity field
+  template <int dim>
+  void
+  ParticleTracking<dim>::setup_background_dofs()
+  {
+    fluid_dh.distribute_dofs(fluid_fe);
+    IndexSet locally_owned_dofs = fluid_dh.locally_owned_dofs();
+    IndexSet locally_relevant_dofs;
+    DoFTools::extract_locally_relevant_dofs(fluid_dh, locally_relevant_dofs);
+
+    field_owned.reinit(locally_owned_dofs, mpi_communicator);
+    field_relevant.reinit(locally_owned_dofs,
+                          locally_relevant_dofs,
+                          mpi_communicator);
+
+    pcout << "Number of degrees of freedom in background grid: "
+          << fluid_dh.n_dofs() << std::endl;
+  }
+
+  template <int dim>
+  void
+  ParticleTracking<dim>::interpolate_function_to_field()
+  {
+    const MappingQ<dim> mapping(fluid_fe.degree);
+
+    VectorTools::interpolate(mapping, fluid_dh, velocity, field_owned);
+    field_relevant = field_owned;
+  }
+
+  template <int dim>
+  void
+  ParticleTracking<dim>::euler_interpolated(double dt)
+  {
+    std::vector<types::global_dof_index> dof_indices(fluid_fe.dofs_per_cell);
+
+    Tensor<1, dim> particle_velocity;
+
+    auto particle = particle_handler.begin();
+    while (particle != particle_handler.end())
+      {
+        const auto &cell =
+          particle->get_surrounding_cell(background_triangulation);
+        const auto &dh_cell =
+          typename DoFHandler<dim>::cell_iterator(*cell, &fluid_dh);
+        dh_cell->get_dof_indices(dof_indices);
+        const auto pic = particle_handler.particles_in_cell(cell);
+
+        for (unsigned int i = 0; particle != pic.end(); ++particle, ++i)
+          {
+            const auto &reference_location = particle->get_reference_location();
+            particle_velocity              = 0.;
+            for (unsigned int j = 0; j < fluid_fe.dofs_per_cell; ++j)
+              {
+                const auto comp_j = fluid_fe.system_to_component_index(j);
+
+                particle_velocity[comp_j.first] +=
+                  fluid_fe.shape_value(j, reference_location) *
+                  field_relevant(dof_indices[j]);
```

Comment:
I would read the values first from the global vector into a cell-local vector and only operate with `j` here.